File: backend/app/schemas/passenger.py
from typing import Optional

from app.schemas.flight_seat import FlightSeatsBase
from .base import SchemaModel
from datetime import datetime
from pydantic import Field


class PassengerBase(SchemaModel):
    citizen_id: str = Field(..., description="Unique citizen identifier")
    passport_number: Optional[str] = Field(None, description="Optional passport number")
    gender: bool = Field(..., description="TRUE for male, FALSE for female")
    phone_number: str = Field(..., description="Passenger's phone number")
    first_name: str = Field(..., description="Passenger's first name")
    last_name: str = Field(..., description="Passenger's last name")
    nationality: str = Field(..., description="Passenger's nationality")
    date_of_birth: datetime = Field(..., description="Passenger's date of birth")
    seat_row: int = Field(..., description="Seat row number")
    seat_col: str = Field(..., description="Seat column identifier (e.g., A, B, C)")


class PassengerCreate(PassengerBase):
    booking_id: str = Field(..., description="Associated booking ID")


class PassengerRead(PassengerBase):
    passenger_id: int


class PassengerUpdate(SchemaModel):
    passport_number: Optional[str] = Field(None, description="Optional passport number")
    gender: bool = Field(..., description="TRUE for male, FALSE for female")
    phone_number: str = Field(..., description="Passenger's phone number")
    first_name: str = Field(..., description="Passenger's first name")
    last_name: str = Field(..., description="Passenger's last name")
    nationality: str = Field(..., description="Passenger's nationality")
    date_of_birth: datetime = Field(..., description="Passenger's date of birth")
    seat_row: int = Field(..., description="Seat row number")
    seat_col: str = Field(..., description="Seat column identifier (e.g., A, B, C)")

File: frontend/src/types/passenger.ts
export interface Passenger {
  citizen_id: string;
  passport_number: string;
  gender: string;
  phone_number: string;
  first_name: string;
  last_name: string;
  nationality: string;
  date_of_birth: string;
  seat_row: number;
  seat_col: string;
}

export let defaultPassenger: Passenger = {
  citizen_id: null as unknown as string,
  passport_number: null as unknown as string,
  gender: null as unknown as string,
  phone_number: null as unknown as string,
  first_name: null as unknown as string,
  last_name: null as unknown as string,
  nationality: null as unknown as string,
  date_of_birth: null as unknown as string,
  seat_row: null as unknown as number,
  seat_col: null as unknown as string,
};

export function updatePassengerSeat(
  passenger: Passenger,
  seatID: string
): Passenger {
  return {
    ...passenger, // Preserve other properties
    seat_row: Number(seatID.slice(0, -1)), // Update seat_row
    seat_col: seatID.slice(-1), // Update seat_col
  };
}
